tests/unit/scheduler.c

```diff
@@ -72,331 +56,257 @@ int verify_schedule(nccl_net_ofi_schedule_t *schedule, nccl_net_ofi_schedule_t *
 	return ret;
 }
 
-int test_multiplexing_schedule()
+int create_ref_schedule(nccl_net_ofi_schedule_t **schedule, int num_xfer_infos)
 {
-	nccl_net_ofi_schedule_t *schedule = NULL;
-	nccl_net_ofi_schedule_t *ref_schedule = (nccl_net_ofi_schedule_t *)malloc(
-		sizeof(nccl_net_ofi_schedule_t) + 3 * sizeof(nccl_net_ofi_xfer_info_t));
-	if (!ref_schedule) {
-		NCCL_OFI_WARN("Could not allocate schedule");
-		return -ENOMEM;
-	}
-	size_t size;
-	int num_rails;
-	size_t align;
 	int ret = 0;
+	*schedule = (nccl_net_ofi_schedule_t *)malloc(sizeof(nccl_net_ofi_xfer_info_t) +
```

Comment:
went to fix some warnings that this introduced under C++, came across this. This is wrong and something we missed in review.

You meant to do:

```
*schedule = (nccl_net_ofi_schedule_t*)malloc(
                sizeof(nccl_net_ofi_schedule_t) + 
                (num_xfer_infos * sizeof(nccl_net_ofi_xfer_info_t)));
```

Looks to me like this was just a typo, in which case, you can ignore the rest of this comment... but if you were confused and afraid to ask and just copy-pasted from somewhere, I wouldn't blame you, because unless you've come across flexible arrays at some point in the past it's a really weird hack that deserves some explanation.

Structs with flexible array members look like this:

```
struct foo {
  int a;
  int b;
  uint64_t flex[];
}
```

Or like this:
```
struct foo {
  int a;
  int b;
  uint64_t flex[0];
}
```


The way this works is that the `sizeof()` operator on this struct does not account for the last member (the flexible member) in any way:

```
static_assert(sizeof(struct foo) == (sizeof(int)*2));
```

This means that if you call:

```
struct foo *my_foo = (struct foo*)malloc(sizeof(struct foo))`
```

You're only allocating space for `a` and `b`, but any access of `my_foo->flex` would be illegal. But the power of this is that if you do:

```
struct foo *my_foo = (struct foo*)malloc(
        sizeof(struct foo) + 
        (nelems * sizeof(uint64_t)))
```

You can then access up to `my_foo->flex[nelems-1]` legally and in a well-defined way.

Some things to watch out for: the flexible array member must always be the last member, and from that rule it follows that there can only ever be one flexible array member in any given struct, and it also follows that if there is some other struct `bar` containing a `struct foo`, that member must come last (and recursively for any struct containing `bar`, and so on):

```
struct bar {
  // struct foo my_foo; // not ok
  int c;
  int d;
  struct foo my_foo; // ok
};
```

Reply:
Thanks for the detailed explanation on flexible array members Nick.
I think I did copy from a previous reference in the code. The explanation really helped.
I believe in this case we should be good as rail_xfer_infos is the last member of the struct nccl_net_ofi_schedule
```
typedef struct nccl_net_ofi_schedule {
	/* Number of transfer information entries set by the scheduler */
	size_t num_xfer_infos;

	/* Array of transfer information structs. The array has at
	 * least 'num_xfer_infos' entries. */
	nccl_net_ofi_xfer_info_t rail_xfer_infos[];
} nccl_net_ofi_schedule_t;
```

I have posted another PR to fix the typo in the function declaration
- https://github.com/aws/aws-ofi-nccl/pull/649

Thank you